[Experimental] Exercise storage access quirks for organizations

sysrqb opened the pull request:
#### 9d07c0773a7ee07ff4814896959591926a25e7c2
<pre>
[Experimental] Exercise storage access quirks for organizations
<a href="https://bugs.webkit.org/show_bug.cgi?id=265800">https://bugs.webkit.org/show_bug.cgi?id=265800</a>
<a href="https://rdar.apple.com/119141675">rdar://119141675</a>

Reviewed by Alex Christensen.

Some organizations operate multiple sites and require complicated interactions
across those sites, such as for signing in. This patch uses a new experimental
quirking mechanism that defines a structure for requesting storage access on
multiple subframe sites for different top sites.

The quirks are updatable, but that is handled in a separate patch. This patch
concentrates on requesting storage access for a subframe before the iframe is
loaded. The timing is important because some sites require storage access
(e.g., cookie access) when requesting the document. This patch accomplishes
that goal by caching the set of relevant quirked subframe sites in the main
document&apos;s Quirks. When an iframe is created we look for the iframe&apos;s site
within the Quirk set and we only initiate the storage access prompt if the
subframe&apos;s site matches a site in the quirk set.

This patch introduces a new Storage Access prompt on Cocoa platforms.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/html/HTMLFrameElementBase.cpp:
(WebCore::HTMLFrameElementBase::openURL):
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::loadStorageAccessQuirksIfNeeded):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::checkLoadCompleteForThisFrame):
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::triggerOptionalStorageAccessIframeQuirk const):
* Source/WebCore/page/Quirks.h:
(WebCore::Quirks::setSubFrameDomainsForStorageAccessQuirk):
(WebCore::Quirks::subFrameDomainsForStorageAccessQuirk const):
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::updatableStorageAccessPromptQuirks):
(WebCore::NetworkStorageSession::updateStorageAccessPromptQuirks):
(WebCore::NetworkStorageSession::loginDomainMatchesRequestingDomain):
(WebCore::NetworkStorageSession::storageAccessQuirkForTopFrameDomain):
(WebCore::NetworkStorageSession::storageAccessQuirkForDomainPair):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::grantStorageAccess):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::requestStorageAccess):
(WebKit::WebResourceLoadStatisticsStore::requestStorageAccessEphemeral):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::storageAccessQuirkForTopFrameDomain):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::updateStorageAccessPromptQuirks):
(WebKit::NetworkProcess::grantStorageAccessForTesting):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/Shared/WebProcessCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/APINavigationClient.h:
(API::NavigationClient::didPromptForStorageAccess):
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageUIClient):
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _grantStorageAccessForTesting:withSubFrameDomains:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::setNavigationDelegate):
(WebKit::NavigationState::NavigationClient::didPromptForStorageAccess):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.h:
* Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
(WebKit::presentStorageAccessAlert):
(WebKit::presentStorageAccessAlertQuirk):
(WebKit::presentStorageAccessAlertSSOQuirk):
(WebKit::displayStorageAccessAlert):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::grantStorageAccessForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::loadStorageAccessQuirksIfNeeded):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
(WebKit::WebProcess::haveStorageAccessQuirksForDomain):
(WebKit::WebProcess::updateDomainsWithStorageAccessQuirks):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
(TEST):
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.mm:
(-[TestNavigationDelegate _webView:didPromptForStorageAccess:forSubFrameDomain:forQuirk:]):

Canonical link: <a href="https://commits.webkit.org/271821@main">https://commits.webkit.org/271821@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5792ead1982b6f2970639e8bc9ca4222118a679a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29774 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8434 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31076 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32275 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26918 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30396 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10601 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5701 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26940 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30053 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7048 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25388 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6006 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6168 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26468 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33612 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27144 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26892 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32350 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6097 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4292 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30132 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7844 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7059 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6850 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6628 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/en.lproj/Localizable.strings:
```diff
@@ -424,6 +424,15 @@
 "Do you want to allow “%@” to use cookies and website data while browsing “%@”?" = "Do you want to allow “%@” to use cookies and website data while browsing “%@”?";
 
+/* Message for requesting cross-site cookie and website data access. */
+"Are you logging in to this website and do you want to allow other %@ sites access to your website data while browsing the websites listed below?" = "Are you logging in to this website and do you want to allow other %@ sites access to your website data while browsing the websites listed below?";
```
comma before and?

achristensen07 commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -428,10 +428,46 @@ const HashMap<RegistrableDomain, HashSet<RegistrableDomain>>& NetworkStorageSess
 
+HashSet<OrganizationStorageAccessPromptQuirk>& NetworkStorageSession::mutableDynamicStorageAccessQuirks()
+{
+    static NeverDestroyed<HashSet<OrganizationStorageAccessPromptQuirk>> set;
```
It might be worth adding ASSERT(RunLoop::isMain()) here.

achristensen07 commented on Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
```diff
@@ -742,6 +743,17 @@ void WebLocalFrameLoaderClient::dispatchDidFinishLoad()
+    if (!m_frame->coreLocalFrame() || !m_frame->coreLocalFrame()->isMainFrame() || !m_frame->coreLocalFrame()->document())
+        return;
+    auto* document = m_frame->coreLocalFrame()->document();
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::StorageAccessQuirkForTopFrameDomain(RegistrableDomain { document->url() }), [document, weakDocument = WeakPtr { *document }](Vector<RegistrableDomain>&& domains) {
```
Could we just capture weakDocument = WeakPtr { document } without also capturing the raw pointer to the document and without dereferencing an unchecked but probably non-null pointer?

Also, is there a way we can do a quick check in the web content process to determine whether we need to send this message?  I imagine the small minority of frame loads will need this.

This could also be a separate FrameLoaderClient call because it doesn't really have anything to do with the finishing of the frame load.

sysrqb replied:
We already restrict this to the main frame. We can further restrict it to main frames where we have quirks. I'll pre-load that set into the web process.

I can move it this call into a separate method, the important timing is that this is called after the document is loaded and before any iframes are created.

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -96,6 +123,14 @@ void displayStorageAccessAlert(WKWebView *webView, NSString *alertTitle, NSStrin
 #else
     auto alert = WebKit::createUIAlertController(alertTitle, informativeText);
 
+    if (accessoryTextList) {
```
I don't think this null check is needed.  ObjC kind of does it for you.

sysrqb replied:
Maybe I want to check something else here. I'm passing nil from two of the functions, and I'm passing a list from the other function. I only want to create the accessory view when the list is present.

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -447,6 +447,13 @@
     }
 
+    if (organizationStorageAccessPromptQuirk) {
+#if !PLATFORM(WATCHOS) && !PLATFORM(APPLETV)
```
This should probably enclose the whole if and return.
Also, can we give it a name, like #if !HAVE(whatever_is_needed_here)

sysrqb replied:
I was following the existing pattern in that function, seems like we can clean that up a bit.

achristensen07 commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -35,13 +35,21 @@
+static HashSet<OrganizationStorageAccessPromptQuirk>& updatableStorageAccessPromptQuirks()
+{
+    ASSERT(RunLoop::isMain());
+    static MainThreadNeverDestroyed<HashSet<OrganizationStorageAccessPromptQuirk>> set;
```
Can this be a member of the NetworkStorageSession rather than a process-global set?

sysrqb replied:
I was preferring a process-global set because it's we only need a single copy of it for the process.

achristensen07 replied:
This should probably have a FIXME comment indicating our intent to move it to something owned by an instance of an object, probably NetworkStorageSession.

johnwilander commented on Source/WebCore/en.lproj/Localizable.strings:
```diff
@@ -424,6 +424,15 @@
 "Do you want to allow “%@” to use cookies and website data while browsing “%@”?" = "Do you want to allow “%@” to use cookies and website data while browsing “%@”?";
 
+/* Message for requesting cross-site cookie and website data access. */
+"Are you logging in to this website, and do you want to allow other %@ sites access to your website data while browsing the websites listed below?" = "Are you logging in to this website, and do you want to allow other %@ sites access to your website data while browsing the websites listed below?";
```
We probably need to reconcile "your website data" here with "their cookies" further down.

johnwilander commented on Source/WebCore/html/HTMLFrameElementBase.cpp:
```diff
@@ -97,13 +97,21 @@ void HTMLFrameElementBase::openURL(LockHistory lockHistory, LockBackForwardList
-        return;
-    }
+    auto completeURL = document().completeURL(m_frameURL);
+    auto finishOpeningURL = [this, weakThis = WeakPtr { *this }, frameName, lockHistory, lockBackForwardList, parentFrame = WTFMove(parentFrame), completeURL] {
```
Is the word "finish" right in this variable name? It sounds like a verb and thus like a function name. Is this like "the URL that we should ultimately open in the sub frame after checking for quirks"? If so, I'd prefer `frameURLToLoadAfterHandlingQuirks` or similar.

sysrqb replied:
This is a lambda that is a completionHandler for when we decide if we should trigger the quirk or finish loading the URL.

johnwilander commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1129,6 +1130,25 @@ Quirks::StorageAccessResult Quirks::requestStorageAccessAndHandleClick(Completio
+void Quirks::triggerOptionalStorageAccessIframeQuirk(const URL& frameURL, CompletionHandler<void()>&& completionHandler) const
+{
+    if (m_document) {
+        if (m_document->frame() && !m_document->frame()->isMainFrame() && m_document->frame()->mainFrame().opener()) {
```
I would break this out into a convenience function to make this code more readable.

johnwilander replied:
Maybe something like `Optional<Frame&> getOpener()`.

johnwilander replied:
Also, do we need a special case for opaque origins here?

achristensen07 replied:
Optional<Frame&> is a Frame*

sysrqb replied:
This will be cleaned up so it correctly supports site isolation, too.

johnwilander commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1129,6 +1130,25 @@ Quirks::StorageAccessResult Quirks::requestStorageAccessAndHandleClick(Completio
+{
+    if (m_document) {
+        if (m_document->frame() && !m_document->frame()->isMainFrame() && m_document->frame()->mainFrame().opener()) {
+            auto& opener = m_document->frame()->mainFrame();
```
… which would be called here and then checked.

achristensen07 replied:
This variable name should be changed to mainFrame.  This is not the opener.

johnwilander commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1129,6 +1130,25 @@ Quirks::StorageAccessResult Quirks::requestStorageAccessAndHandleClick(Completio
+                return;
+            }
+        }
+        if (subFrameDomainsForStorageAccessQuirk().contains(RegistrableDomain { frameURL })) {
```
Opaque origins may cause unexpected behavior here.

johnwilander commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2016-2018 Apple Inc.  All rights reserved.
+ * Copyright (C) 2016-2023 Apple Inc.  All rights reserved.
```
Always good to see clean-up of these.

johnwilander commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -458,6 +474,32 @@ std::optional<RegistrableDomain> NetworkStorageSession::findAdditionalLoginDomai
+Vector<RegistrableDomain> NetworkStorageSession::storageAccessQuirkForTopFrameDomain(const TopFrameDomain& topDomain)
+{
+    for (auto&& quirk : updatableStorageAccessPromptQuirks()) {
+        auto& domainPairings = quirk.domainPairings;
```
This and the three following lines are repeated in the next function. Could they be broken out into a convenience function which would improve readability?

johnwilander commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -432,7 +432,9 @@ void WebResourceLoadStatisticsStore::requestStorageAccess(RegistrableDomain&& su
             };
-            m_networkSession->networkProcess().parentProcessConnection()->sendWithAsyncReply(Messages::NetworkProcessProxy::RequestStorageAccessConfirm(webPageProxyID, frameID, subFrameDomain, topFrameDomain), WTFMove(requestConfirmationCompletionHandler));
+
+            auto quirk = NetworkStorageSession::storageAccessQuirkForDomainPair(topFrameDomain, subFrameDomain);
```
I would call it `storageAccessQuirk` for readability on the next line.

johnwilander commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -471,7 +473,8 @@ void WebResourceLoadStatisticsStore::requestStorageAccessEphemeral(const Registr
     };
 
-    m_networkSession->networkProcess().parentProcessConnection()->sendWithAsyncReply(Messages::NetworkProcessProxy::RequestStorageAccessConfirm(webPageProxyID, frameID, subFrameDomain, topFrameDomain), WTFMove(requestConfirmationCompletionHandler));
+    auto quirk = NetworkStorageSession::storageAccessQuirkForDomainPair(topFrameDomain, subFrameDomain);
```
Ditto.

johnwilander commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1311,6 +1312,16 @@ void NetworkProcess::resetCrossSiteLoadsWithLinkDecorationForTesting(PAL::Sessio
     completionHandler();
 }
 
+void NetworkProcess::grantStorageAccessForTesting(PAL::SessionID sessionID, Vector<RegistrableDomain>&& subFrameDomains, RegistrableDomain&& topFrameDomain, CompletionHandler<void(void)>&& completionHandler)
```
Is this only for our automated testing? If so, please lock it down to the domains we use in testing. That way a rogue WebKit client can't get to this function to override cookie blocking.

achristensen07 replied:
Alternatively we could do something like [EnabledIf='allowTestOnlyIPC()']

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -455,6 +455,14 @@ static float toNSURLSessionTaskPriority(WebCore::ResourceLoadPriority priority)
         request.setFirstPartyForCookies(request.url());
+    else {
+        WebCore::RegistrableDomain firstPartyDomain { request.firstPartyForCookies() };
+        if (auto* storageSession = m_session->networkStorageSession(); storageSession
```
Please break out the two checks – redirect and request – into two local variables with good names. That will make this if statement much more readable.

johnwilander commented on Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
```diff
@@ -948,6 +949,19 @@ static bool isUnsupportedWebExtensionNavigation(API::NavigationAction& navigatio
+        return;
+
+    if (m_navigationState->m_navigationDelegateMethods.webViewDidPromptForStorageAccessForSubFrameDomainForQuirk)
+        [static_cast<id<WKNavigationDelegatePrivate>>(navigationDelegate) _webView:m_navigationState->webView().get() didPromptForStorageAccess:topFrameDomain forSubFrameDomain:subFrameDomain forQuirk:hasQuirk];
```
Are we guaranteed to always have a webView here? And that the lifetime of it can not terminate before we're done with this?

johnwilander commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -1415,3 +1415,191 @@ - (void)webView:(WKWebView *)webView stopURLSchemeTask:(id <WKURLSchemeTask>)tas
     TestWebKitAPI::Util::run(&done);
 }
+
+TEST(ResourceLoadStatistics, StorageAccessPromptSiteWithQuirk)
```
Without reading all the details in this test, does it also test negatively? I.e. does it test that storage is not granted to a domain that's not in the set?

johnwilander reviewed (commented):
We should call this feature experimental since it is. Both as a prefix in the bug title and as a line in the change log entry.

Missing a plural 's' at the end of this change log statement: "This patch introduces a new Storage Access prompt on Cocoa platform."

My code comments are mostly on readability and asking questions on lifetime of some data. As for readability and naming, if you're short on time, I'm fine with those landing in a risk-free follow-up patch.

achristensen07 commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1129,6 +1130,25 @@ Quirks::StorageAccessResult Quirks::requestStorageAccessAndHandleClick(Completio
+    if (m_document) {
+        if (m_document->frame() && !m_document->frame()->isMainFrame() && m_document->frame()->mainFrame().opener()) {
+            auto& opener = m_document->frame()->mainFrame();
+            if (auto* localFrame = dynamicDowncast<LocalFrame>(opener); localFrame && localFrame->document()) {
```
localMainFrame

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1311,6 +1312,19 @@ void NetworkProcess::resetCrossSiteLoadsWithLinkDecorationForTesting(PAL::Sessio
+{
+    HashSet allowedDomains { "site1.example"_str, "site2.example"_str, "site3.example"_str, "site4.example"_str };
+    if (!allowedDomains.contains(topFrameDomain.string()))
+        completionHandler();
```
This needs to return so we don't call null.

achristensen07 commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -455,6 +455,16 @@ static float toNSURLSessionTaskPriority(WebCore::ResourceLoadPriority priority)
+        if (auto* storageSession = m_session->networkStorageSession()) {
+            bool didPreviousRequestHaveStorageAccess = storageSession->hasStorageAccess(WebCore::RegistrableDomain { redirectResponse.url() }, firstPartyDomain, m_frameID, m_pageID);
+            bool doesRequestHaveStorageAccess = storageSession->hasStorageAccess(WebCore::RegistrableDomain { request.url() }, firstPartyDomain, m_frameID, m_pageID);
+            WTFLogAlways("didPreviousRequestHaveStorageAccess: %d, doesRequestHaveStorageAccess: %d", didPreviousRequestHaveStorageAccess, doesRequestHaveStorageAccess);
```
Please remove.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -650,6 +650,14 @@ - (void)_setStorageAccessPromptQuirkForTesting:(NSString *)topFrameDomain withSu
+- (void)_grantStorageAccessForTesting:(NSString *)topFrameDomain withSubFrameDomains:(NSArray<NSString *> *)subFrameDomains completionHandler:(void(^)(void))completionHandler
+{
+    if (!_websiteDataStore->isPersistent())
+        return;
```
call completionHandler?

achristensen07 commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2113,6 +2115,17 @@ void WebProcess::sendResourceLoadStatisticsDataImmediately(CompletionHandler<voi
+void WebProcess::updateDomainsWithStorageAccessQuirks(HashSet<WebCore::RegistrableDomain>&& domainsWithStorageAccessQuirks)
+{
+    for (auto&& domain : domainsWithStorageAccessQuirks)
+        m_domainsWithStorageAccessQuirks.add(WTFMove(domain));
```
Do we want to clear them first?

sysrqb committed:
[Experimental] Exercise storage access quirks for organizations
https://bugs.webkit.org/show_bug.cgi?id=265800
rdar://119141675

Reviewed by Alex Christensen.

Some organizations operate multiple sites and require complicated interactions
across those sites, such as for signing in. This patch uses a new experimental
quirking mechanism that defines a structure for requesting storage access on
multiple subframe sites for different top sites.

The quirks are updatable, but that is handled in a separate patch. This patch
concentrates on requesting storage access for a subframe before the iframe is
loaded. The timing is important because some sites require storage access
(e.g., cookie access) when requesting the document. This patch accomplishes
that goal by caching the set of relevant quirked subframe sites in the main
document's Quirks. When an iframe is created we look for the iframe's site
within the Quirk set and we only initiate the storage access prompt if the
subframe's site matches a site in the quirk set.

This patch introduces a new Storage Access prompt on Cocoa platforms.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/html/HTMLFrameElementBase.cpp:
(WebCore::HTMLFrameElementBase::openURL):
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::loadStorageAccessQuirksIfNeeded):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::checkLoadCompleteForThisFrame):
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::triggerOptionalStorageAccessIframeQuirk const):
* Source/WebCore/page/Quirks.h:
(WebCore::Quirks::setSubFrameDomainsForStorageAccessQuirk):
(WebCore::Quirks::subFrameDomainsForStorageAccessQuirk const):
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::updatableStorageAccessPromptQuirks):
(WebCore::NetworkStorageSession::updateStorageAccessPromptQuirks):
(WebCore::NetworkStorageSession::loginDomainMatchesRequestingDomain):
(WebCore::NetworkStorageSession::storageAccessQuirkForTopFrameDomain):
(WebCore::NetworkStorageSession::storageAccessQuirkForDomainPair):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::grantStorageAccess):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::requestStorageAccess):
(WebKit::WebResourceLoadStatisticsStore::requestStorageAccessEphemeral):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::storageAccessQuirkForTopFrameDomain):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::updateStorageAccessPromptQuirks):
(WebKit::NetworkProcess::grantStorageAccessForTesting):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/Shared/WebProcessCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/APINavigationClient.h:
(API::NavigationClient::didPromptForStorageAccess):
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageUIClient):
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _grantStorageAccessForTesting:withSubFrameDomains:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::setNavigationDelegate):
(WebKit::NavigationState::NavigationClient::didPromptForStorageAccess):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.h:
* Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
(WebKit::presentStorageAccessAlert):
(WebKit::presentStorageAccessAlertQuirk):
(WebKit::presentStorageAccessAlertSSOQuirk):
(WebKit::displayStorageAccessAlert):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestStorageAccessConfirm):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::grantStorageAccessForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::loadStorageAccessQuirksIfNeeded):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
(WebKit::WebProcess::haveStorageAccessQuirksForDomain):
(WebKit::WebProcess::updateDomainsWithStorageAccessQuirks):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
(TEST):
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.mm:
(-[TestNavigationDelegate _webView:didPromptForStorageAccess:forSubFrameDomain:forQuirk:]):

Canonical link: https://commits.webkit.org/271821@main
